### PR TITLE
Allow file fields in question groups

### PR DESCRIPTION
**Pull request by oliverroick**
### Proposed changes in this pull request

The PR addresses the need to define resource fields (images, videos) inside question groups for locations, relationships and parties. 

Changes added include:

- Filter out resource-related fields in `questionnaires.managers.create_attrs_schema` when creating `Attribute` instances. Resources are not part of attribute schemas and should be omitted. We identify resource files using the field-name pattern required to specify resources; i.e. `{model_type}_resource`.
- Omit resources from attribute validation in `ModelHelper. _get_attributes` ([LL429-432](https://github.com/Cadasta/cadasta-platform/blob/656ada6280cf41ae558aa654633cfc2ca59d931d/cadasta/xforms/mixins/model_helper.py#L429-L432))
- Extract resources from the GeoODK submission so they can be attached to the correct model instances. In `ModelHelper. _get_resource_names` ([LL464-467](https://github.com/Cadasta/cadasta-platform/blob/656ada6280cf41ae558aa654633cfc2ca59d931d/cadasta/xforms/mixins/model_helper.py#L464-L467)), we're drilling down into the attribute groups, identify resources using the same pattern described before and attach them to the list of resources. 

### When should this PR be merged

If possible with 1.15.0 as this will make running one of Katrina's projects easier. This should be tested using GeoODK the form that is used in the project.

### Risks

Low.

### Follow-up actions

None.

#### General

**Is this PR explained thoroughly?** All code changes must be accounted for in the PR description. 

- [ ] Review 1
- [ ] Review 2

**Is the PR labeled correctly?** It should have the `migration` label if a new migration is added. 

- [ ] Review 1
- [ ] Review 2

**Is the risk level assessment sufficient?** The risks section should contain all risks that might be introduced with the PR and which actions we need to take to mitigate these risks. Possible risks are database migrations, new libraries that need to be installed or changes to deployment scripts.

- [ ] Review 1
- [ ] Review 2

#### Functionality

**Are all requirements met?** Compare implemented functionality with the requirements specification.

- [ ] Review 1
- [ ] Review 2

**Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled. 

- [ ] Review 1
- [ ] Review 2

#### Code

**Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.

- [ ] Review 1
- [ ] Review 2

**Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests. 

- [ ] Review 1
- [ ] Review 2

**Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/). 

- [ ] Review 1
- [ ] Review 2

**Is the code documented sufficiently?** Large and complex classes, functions or methods must be annotated with comments following our [code-style guidelines](https://devwiki.corp.cadasta.org/Contributing%20Style%20Guide#documentation-and-comments).

- [ ] Review 1
- [ ] Review 2

**Has the scalability of this change been evaluated?**

- [ ] Review 1
- [ ] Review 2

**Is there a maintenance plan in place?**

- [ ] Review 1
- [ ] Review 2

#### Tests

**Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components. 

- [ ] Review 1
- [ ] Review 2

**If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.

- [ ] Review 1
- [ ] Review 2

**If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

- [ ] Review 1
- [ ] Review 2

#### Security

**Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**

- [ ] Review 1
- [ ] Review 2

**Are all UI and API inputs run through forms or serializers?** 

- [ ] Review 1
- [ ] Review 2

**Are all external inputs validated and sanitized appropriately?**

- [ ] Review 1
- [ ] Review 2

**Does all branching logic have a default case?**

- [ ] Review 1
- [ ] Review 2

**Does this solution handle outliers and edge cases gracefully?**

- [ ] Review 1
- [ ] Review 2

**Are all external communications secured and restricted to SSL?**

- [ ] Review 1
- [ ] Review 2

#### Documentation

**Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes must be documented in the [Cadasta Platform Documentation](https://github.com/Cadasta/cadasta-docs).

- [ ] Review 1
- [ ] Review 2

**Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented in the [API docs](https://github.com/Cadasta/api-docs).

- [ ] Review 1
- [ ] Review 2

**Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki. 

- [ ] Review 1
- [ ] Review 2

### CODE REVIEW

**Comment by MappingKat** [comment]
FYI @seav I got an error when trying to upload a questionnaire with images inside of the location attributes... I havent look at it in detail because im tired and its late. Let me know if you know what this could be!
[Public_Space_Assessment_Inventory-images.xlsx](https://github.com/Cadasta/cadasta-platform/files/1577838/Public_Space_Assessment_Inventory-images.xlsx)
screen shot 2017-12-20 at 10 43 03 pm

**Comment by MappingKat** [comment]
Hm. When I load that questionnaire on staging it tells me: 
screen shot 2017-12-20 at 10 48 05 pm

**Comment by seav** [comment]
> Hm. When I load that questionnaire on staging it tells me:

@MappingKat: Staging still does not have this PR. So resource files inside groups are not yet supported on staging.

**Comment by seav** [comment]
I added a commit to fix the bug that I detected. Basically, `ModelHelper._get_resource_files()` was intended to construct a flat list of resource filenames that are attached to either parties, locations, and tenure relationships. The changes in `ModelHelper._get_resource_names()` should have also been applied to `ModelHelper._get_resource_files()`. But I figured that since `ModelHelper._get_resource_names()` was already extracting these filenames, then we don't actually need `ModelHelper._get_resource_files()` and we can just flatten the extracted filenames already. This simplified the code a bit.

I was able to confirm that this fix works using ODK. I also checked that resource attachment still works for resource questionnaire items outside of groups.

File: cadasta/questionnaires/managers.py [hunk 1]
```diff
@@ -106,6 +106,13 @@ def create_attrs_schema(project, question_group_dict, attr_type_ids,
         field = {}
         field['name'] = child.get('name')
         field['long_name'] = child.get('label')
+
+        field['name'] = child.get('name')
```
Comment (alukach) [inline cadasta/questionnaires/managers.py]: This is a repeat of line 107.

**Pull request merged**